File: controllers/search.controller.js
const MenuItem = require('../models/MenuItem');
const Vendor = require('../models/Vendor');

// GET /api/search?q=biryani
exports.search = async (req, res) => {
  try {
    const query = req.query.q;
    if (!query) {
      return res.status(400).json({ message: 'Search query parameter q is required.' });
    }

    const regex = new RegExp(query, 'i');

    // 1. Search Vendors
    const vendors = await Vendor.find({
      isActive: true,
      isApproved: true,
      $or: [{ name: regex }, { cuisine: regex }, { address: regex }]
    });

    // 2. Search Menu Items
    const foods = await MenuItem.find({
      isAvailable: true,
      $or: [{ name: regex }, { description: regex }, { category: regex }, { tags: { $in: [regex] } }]
    }).populate('vendor', 'name cuisine address rating');

    res.json({ vendors, foods });
  } catch (err) {
    res.status(500).json({ message: err.message });
  }
};

// GET /api/foods/filter?calories=500&category=healthy&city=Vehari
exports.filterFoods = async (req, res) => {
  try {
    const { calories, category, city } = req.query;
    const foodQuery = { isAvailable: true };

    // 1. Calorie Filter
    if (calories) {
      foodQuery.calories = { $lte: Number(calories) };
    }

    // 2. Category Filter
    if (category) {
      const catRegex = new RegExp(category, 'i');
      foodQuery.$or = [
        { category: catRegex },
        { tags: { $in: [category.toLowerCase()] } }
      ];
    }

    // 3. City/Vendor Filter
    if (city) {
      const cityRegex = new RegExp(city, 'i');
      const vendorsInCity = await Vendor.find({
        isActive: true,
        address: cityRegex
      }).select('_id');

      const vendorIds = vendorsInCity.map(v => v._id);
      foodQuery.vendor = { $in: vendorIds };
    }

    const foods = await MenuItem.find(foodQuery)
      .populate('vendor', 'name cuisine address rating');

    res.json(foods);
  } catch (err) {
    res.status(500).json({ message: err.message });
  }
};

File: controllers/search_controller.py
from fastapi import HTTPException, status
from sqlalchemy import select, or_, and_
from database import AsyncSessionLocal, serialize_doc
from orm_models import Vendor, Food

async def search_query(q: str):
    if not q:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Search query parameter 'q' is required"
        )

    try:
        async with AsyncSessionLocal() as session:
            pattern = f"%{q}%"

            # 1. Search Vendors
            v_res = await session.execute(
                select(Vendor).where(
                    and_(
                        Vendor.is_active == True,
                        or_(
                            Vendor.name.ilike(pattern),
                            Vendor.cuisine.ilike(pattern),
                            Vendor.category.ilike(pattern),
                            Vendor.address.ilike(pattern),
                            Vendor.city.ilike(pattern)
                        )
                    )
                ).limit(50)
            )
            vendors = v_res.scalars().all()

            # 2. Search Foods
            f_res = await session.execute(
                select(Food).where(
                    and_(
                        Food.is_available == True,
                        or_(
                            Food.name.ilike(pattern),
                            Food.description.ilike(pattern),
                            Food.category.ilike(pattern)
                        )
                    )
                ).limit(50)
            )
            foods = f_res.scalars().all()

            serialized_vendors = serialize_doc(list(vendors))
            for v in serialized_vendors:
                v.pop("password", None)

            serialized_foods = serialize_doc(list(foods))
            for food in serialized_foods:
                v_id = food.get("vendor_id")
                if v_id:
                    v_doc_res = await session.execute(select(Vendor).where(Vendor.id == str(v_id)))
                    v_doc = v_doc_res.scalar_one_or_none()
                    if v_doc:
                        v_clean = serialize_doc(v_doc)
                        v_clean.pop("password", None)
                        food["vendor"] = v_clean

            return {
                "query": q,
                "vendors": serialized_vendors,
                "foods": serialized_foods
            }

    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Search failed: {str(e)}"
        )
